### ISSUE TITLE
[vim9class] Illegal lookup with upcasting for instance methods

### ISSUE THREAD

**Issue by zzzyxwvut** [vim/vim#15432]
### Steps to reproduce

Upcasting from derived types to their base types is common  
practice: a) library authors don't know and shouldn't care  
about derived type hierarchies based on their API, b) users  
of libraries may follow suit and narrow their derived types  
to some common base API types.

Given the type hierarchy:

```
abstract class A
    abstract def ToString(): string

    def string(): string
        return 'A'
    enddef
endclass

class B extends A
    def ToString(): string
        return this.string()
    enddef

##  def string(): string
##      return 'B'
##  enddef
endclass

class C extends B
    def ToString(): string
        return this.string()
    enddef

    def ToLowerCasedString(): string
        return tolower(this.string())
    enddef

    def string(): string
        return 'C'
    enddef
endclass
```

let's explore two issues.

Here, a `B` instance is instantiated and upcast to `A` at its  
assignment, and from now on `B`'s implementation is accessed  
through `A`'s API.  And the same applies for a `C` instance.  
As can be seen, method dispatch is correctly performed.

```
const b_: A = B.new()
echo 'A' == b_.ToString()
echo 'A' == b_.string()

const c_: A = C.new()
echo 'C' == c_.ToString()
echo 'C' == c_.string()
echo 'c' == (<C>c_).ToLowerCasedString()
```

However, it appears that the upcast type is not enforced.

These should fail with `E1325` because `ToLowerCasedString()`  
is not defined in either `A` or `B` and `c_` was narrowed to  
`A`; but they don't:

```
echo '.... 1 ....'
echo 'c' == c_.ToLowerCasedString()
echo 'c' == (<A>c_).ToLowerCasedString()
echo 'c' == (<B>c_).ToLowerCasedString()
const ca: A = <A>c_
const cb: B = <B>c_
echo 'c' == ca.ToLowerCasedString()
echo 'c' == cb.ToLowerCasedString()
```

These correctly fail for `B`:

```
echo b_.ToLowerCasedString()
echo B.new().ToLowerCasedString()
```
Another upcasting issue involves `abstract` methods of upcast  
types.

These `C` declarations are accepted (both `ToString()` and  
`string()` are defined in `C`):

```
echo 'CCCC' == repeat([C.new()], 4)
    ->reduce((prefix: string, c: C) => prefix .. c.string(), '')

echo 'CCCC' == repeat([C.new()], 4)
    ->reduce((prefix: string, c: C) => prefix .. c.ToString(), '')
```

These `A` declarations are rejected (`E1027`) (`ToString()` is  
`abstract` in `A`):

```
echo '.... 2 ....'

echo 'CCCC' == repeat([C.new()], 4)
    ->reduce((prefix: string, a: A) => prefix .. a.ToString(), '')

def TestA(a: A)
    echo a.ToString()
enddef

TestA(C.new())
```

### Expected behaviour

1) Upcast instances should only have access to members of  
their cast type (users are free to downcast instances to  
their runtime types when necessary).

2) The `abstract` methods of an upcast type should not  
interfere in method dispatch whenever their implementations  
are referenced through this upcast type.

### Version of Vim

v9.1.0659

### Environment

GNU/Linux Debian (testing/trixie)

**Comment by zzzyxwvut** [comment]
Let me clarify the expectation for `string()` in these two  
examples:

```
## E1325, when there is no string() definition in A.
echo 'CCCC' == repeat([C.new()], 4)
    ->reduce((prefix: string, a: A) => prefix .. a.string(), '')

## E1325, when there is no string() definition in A.
echo 'CCCC' == repeat([C.new()], 4)
    ->reduce((prefix: string, b: B) => prefix .. b.string(), '')
```

This syntax can only be valid if `string()` is either `abstract`  
or implemented in the declared type or in any of its base  
types (possibly in an `interface`).  So, `string()` and other  
`:help builtin-object-methods` are and should be treated like  
user-defined methods.

The following _modified_ `A` and `B` are not supported (`E1027`)  
when used with the above `C` for the listed examples:

(1)

```
abstract class A
    abstract def string(): string
endclass

abstract class B extends A
endclass
```

(2)

```
interface I
    def string(): string
endinterface

abstract class A implements I
    abstract def string(): string
endclass

abstract class B extends A
    abstract def string(): string
endclass
```

The workaround for lambda expression signatures is to  
declare a type where `string()` is implemented.

**Comment by zzzyxwvut** [comment]
Regular functions also fail (`E1027`) with the (2) example  
types.

```
def TestI(i: I)
    echo i.string()
enddef

def TestA(a: A)
    echo a.string()
enddef

def TestB(b: B)
    echo b.string()
enddef

TestI(C.new())
TestA(C.new())
TestB(C.new())
```

**Comment by errael** [comment]
I'm intrigued by this issue, finally got some time to take a look. The `E1027` is clearly inappropriate. I haven't looked at all the code examples in detail, but I'm wondering if restrictions on `builtin-object-methods` might clean it all up.

Would these rules be acceptable and not too onerous; and would they fix all these issues?
- Error to declare `builtin-object-method` abstract.
- `builtin-object-method` can only be used as a builtin, and not as a method.

The following works fine
```
vim9script

abstract class A
endclass

class B extends A
    def string(): string
        return 'B'
    enddef
endclass

class C extends B
    def string(): string
        return 'C'
    enddef
endclass

def TestA(a: A)
    echo a->string()
enddef

TestA(B.new())
TestA(C.new())
```

**Comment by errael** [comment]
Wondering about casting with `<{type}>`...

I think it's not what I typically think of as type casting. In particular, the docs say
```
The semantics is that, if needed, a runtime type check is performed.
```
The docs might imply that casting is only for use with something of type `<any>`.

It seems like that since that was written, the meaning of `<type>` has been extended a little. I remember a case where it can be used to get rid of "dictionary required" errors. I guess the type check also changes the type (a little bit anyway).

**Comment by zzzyxwvut** [comment]
> Would these rules be acceptable and not too onerous; and would they fix all these issues?

The `abstract` method problem relates to all user-defined  
methods in general, not just builtin object methods.

> - Error to declare `builtin-object-method` abstract.
> - `builtin-object-method` can only be used as a builtin, and not as a method.

As soon as builtin object methods cease to be valid as _type  
methods_, what are they good for?

**Comment by errael** [comment]
> > * Error to declare `builtin-object-method` abstract.
> > * `builtin-object-method` can only be used as a builtin, and not as a method.
> 
> As soon as builtin object methods cease to be valid as _type methods_, what are they good for?

I'm not sure I entirely understand the question. They are good for
using the class with builtins. There are a variety of places where
vim implicitly converts something to a string. Like
```
echo printf("%s", B.new())
```
Is that a good/useful application of defining `string()` for a class?

What's the relevance of having them available as _type methods_?

You can always put something like
```
def String(): string
    return this->string()
enddef
```
in the abstract base class. Does that solve your problem?

**Comment by errael** [comment]
> The `abstract` method problem relates to all user-defined methods in general, not just builtin object methods.

At first I thought this had to be some kind of regression. So I tried the following on vim9.1. Same failure; wow.
```
vim9script

abstract class A
    abstract def String(): string
endclass

class B extends A
    def String(): string
        return 'B'
    enddef
endclass

var x: A = B.new()
echo x.String()         # <<<<< WORKS

def F()
    echo x.String()     # <<<<< FAILS
enddef
disa F
```
Is it accurate to say
> an abstract method with a return value not handled
> when referenced through a variable of the abstract type
> in a compiled function

**Comment by zzzyxwvut** [comment]
> What's the relevance of having them available as type methods?

Well, the description of `builtin-object-methods` says:

https://github.com/vim/vim/blob/91efcd115e700725b9ebded0f5d7bc0d3fa98d9d/runtime/doc/vim9class.txt#L732-L734

so we need to define an instance (type) method, e.g.

```
vim9script

abstract class Text
    def String(): string
        return this->string()
    enddef
endclass

class Hello extends Text
    def string(): string
        return "hello"
    enddef
endclass
```

before any invocation variant is applicable:

```
var hello = Hello.new()
echo 'hello' == hello.string()
echo 'hello' == hello.String()
echo 'hello' == hello->string()
echo 'hello' == string(hello)
```

**Comment by errael** [comment]
> > What's the relevance of having them available as type methods?
> ...
> so we need to define an instance (type) method, e.g.

Right. Defined as type methods. But they don't have to be invoked as type methods. Invoking them as `obj->string()` is sufficient isn't it?

**Comment by zzzyxwvut** [comment]
Sure.  But there are other user-defined methods that suffer  
from the same `abstract` problem.

**Comment by errael** [comment]
> Sure. But there are other user-defined methods that suffer from the same `abstract` problem.

It was not my intention to say there's no problem, see https://github.com/vim/vim/issues/15432#issuecomment-2272169907.

Did my response to the question
> builtin object methods ... what are they good for?

confuse things?

**Comment by errael** [comment]
I am wondering if saying

> an abstract method with a return value not handled
> when referenced through a variable of the abstract type
> in a compiled function

encompasses the issue or if there are other problems.

**Comment by zzzyxwvut** [comment]
According to my findings, the _second_ problem can be  
summarised as follows:

> a non-abstract method is not handled when referenced through
> an instance of an upcast type that only inherits an abstract
> declaration of this method, in a compiled function

**Comment by errael** [comment]
> According to my findings, the _second_ problem can be summarised as follows:
> 
> > a non-abstract instance method is not handled when
> > referenced through an upcast type that only inherits
> > an abstract declaration of this method, in a compiled
> > function

The following seems to meet this criteria, but it works.
```
vim9script

abstract class A
    abstract def String()
endclass

class B extends A
    def String()
        echo 'B'
    enddef
endclass

var x: A = B.new()

def F()
    x.String()
enddef
F()
```
Looks like having a return value is part of the problem.

**Comment by errael** [comment]
Or maybe there's more to it... (and I'm wondering what is the _first_ problem)

**Comment by zzzyxwvut** [comment]
OK, `abstract` `void` methods are not problematic.

The _first_ problem is outlined in the "Expected behaviour":

> 1. Upcast instances should only have access to members of
> their cast type (users are free to downcast instances to
> their runtime types when necessary).

**Comment by errael** [comment]
> > 1. Upcast instances should only have access to members of
> >    their cast type (users are free to downcast instances to
> >    their runtime types when necessary).

IIUC, this is about `<{type}>`. While I agree with the sentiment, it's not what the docs say. Specifically, neither the value nor the **type** is changed. All it needs to do is a runtime check on the type.

https://github.com/vim/vim/blob/91efcd115e700725b9ebded0f5d7bc0d3fa98d9d/runtime/doc/vim9.txt#L1568-L1571

And there's still some weirdness. Sometimes a cast makes a difference at compile time; but should it? It would be nice if this all got cleaned up.
```
vim9script

class A
endclass

class B extends A
    def F()
    enddef
endclass

def G1(o: any)
    (<A>o).F()  # <<<<< compile time error, F() not in A. But seems like a bug.
                # It could be argued that this should be a runtime check.
                # See G3 below. Something seems wrong/inconsistent.
enddef
#disa G1

def G2(o: A)
    (<B>o).F()  # <<<<< this compiles OK. A CHECKTYPE instruction generated.
enddef
disa G2
G2(B.new())     # OK
#G2(A.new())     # runtime error: expected object<B> but got object<A>

def G3(o: B)
    (<A>o).F()  # <<<<< this compiles OK. A "B" can be cast to an "A"
                # no CHECKTYPE instruction
enddef
disa G3
G2(B.new())     # OK. WEIRD. But doesn't violate the doc/spec.
G2(A.new())     # runtime error: expected object<B> but got object<A>
```

**Comment by zzzyxwvut** [comment]
> https://github.com/vim/vim/blob/91efcd115e700725b9ebded0f5d7bc0d3fa98d9d/runtime/doc/vim9.txt#L1568-L1571

This problem's scope concerns permitted (and sensible) _type_  
transformations of some runtime value.  Neither `number` nor  
`string` are subtypes of each other; therefore, functions  
`string()` and `str2nr()` are mentioned since casting either  
value to another is not supported (although numbers do enjoy  
certain textual leeway in specific contexts: `1 .. "1"`,  
dictionary keys).

On the other hand, the types from [the original example](https://github.com/vim/vim/issues/15432#issue-2447282658) form  
a hierarchy, i.e. `(D) < C < B < A`, and, depending on the  
runtime type, some casts are allowed.  It appears that the  
following upcasts:

```
class D extends C
endclass

var c: C = C.new()
var o: any = <any>c
var a: A = <A>c
var b: B = <B>c
var c2: C = <C>c

# Huh?
var c3: C = <D>a
var c4: C = <D>b
var c5: C = <D>c
var c6: C = <D>o

echo o a b c c2 c3 c4 c5 c6
echo o.ToString()
echo a.ToString()
echo b.ToString()

## E1324.
#echo o .. a .. b .. c .. c2 .. c3 .. c4 .. c5 .. c6
```

do identity conversion from and to the runtime type despite  
what assignment signatures may suggest.

**Comment by zzzyxwvut** [comment]
Looking at [the above example](https://github.com/vim/vim/issues/15432#issuecomment-2272389793):

> ```
> def G1(o: any)
>     (<A>o).F()  # <<<<< compile time error, F() not in A. But seems like a bug.
>                 # It could be argued that this should be a runtime check.
>                 # See G3 below. Something seems wrong/inconsistent.
> enddef
> #disa G1
> ```

This is correct, if upcasts only expose cast type members,  
because `F()` is only declared/defined in `B`.  (For example,  
`A` can be a library-exported type, whereas `B` will be its  
derived type written by a library user with some _new_ methods  
introduced like `F()`.)  In addition, I guess, this implicit rule:

https://github.com/vim/vim/blob/91efcd115e700725b9ebded0f5d7bc0d3fa98d9d/runtime/doc/vim9.txt#L1552-L1562

makes the difference here, actually casting from `any` to `A`.

I find what follows problematic:

> ```
> G2(B.new())     # OK. WEIRD. But doesn't violate the doc/spec.
> ```

from this snippet:

> ```
> def G3(o: B)
>     (<A>o).F()  # <<<<< this compiles OK. A "B" can be cast to an "A"
>                 # no CHECKTYPE instruction
> enddef
> disa G3
> G2(B.new())     # OK. WEIRD. But doesn't violate the doc/spec.
> G2(A.new())     # runtime error: expected object<B> but got object<A>
> ```

It appears the cast is but identity conversion, as if it  
were written `o.F()`:

> ```
> def G3(o: B)
>     o.F()       # <<<<< this compiles OK. A "B" can be cast to an "A"
>                 # no CHECKTYPE instruction
> enddef 
> ```

Trying `(<A><any>o).F()` doesn't help.

**Comment by errael** [comment]
> > https://github.com/vim/vim/blob/91efcd115e700725b9ebded0f5d7bc0d3fa98d9d/runtime/doc/vim9.txt#L1568-L1571
> 
> This problem's scope concerns permitted (and sensible) _type_ transformations of some runtime value. Neither `number` nor `string` are subtypes of each other; therefore, functions `string()` and `str2nr()` are mentioned since casting either value to another is not supported (although numbers do enjoy certain textual leeway in specific contexts: `1 .. "1"`, dictionary keys).

Yeah, the `string()` conversion confuses things; in languages I'm familiar with, a cast doesn't do string/number conversions. (ignoring the things you can do with overloading in c++)

A cast, `<{type}>`,  only does a `TYPECHECK` (if needed). It does not change type, although there seem to be some compiler implications when dealing with `any`.

> On the other hand, the types from [the original example](https://github.com/vim/vim/issues/15432#issue-2447282658) form a hierarchy, i.e. `(D) < C < B < A`, and, depending on the runtime type, some casts are allowed.


It appears that the following upcasts:
> 
> ```
> class D extends C
> endclass
> 
> ...
> 
> do identity conversion from and to the runtime type despite what assignment signatures may suggest.

If you remove all the `<{type}>` from this example, you get the same output. By definition/spec cast doesn't do any conversion.

Ahh, I see one problem. The interpreter, script level, does not always behave the same; those cases are considered bugs (AFAIK).

Putting this code in a function
```vim
def F()
    var c: C = C.new()
    var o: any = c
    var a: A = c
    var b: B = c
    var c2: C = c

    # Huh?
    var c3: C = a
    var c4: C = b
    var c5: C = c
    var c6: C = o

    echo o a b c c2 c3 c4 c5 c6
    echo o.ToString()
    echo a.ToString()
    echo b.ToString()
enddef
F()
```
produces
```
Error detected while compiling /tmp/f.vim[67]..function <SNR>46_F:
line    8:
E1012: Type mismatch; expected object<C> but got object<A>
```

So you're "Huh" case is different in compiled code. Note, the above doesn't have the `<{type}>`

With the original statements, that include the `<{type}>`, in a function, there's the "missing return" issue. Using the patch #15441. There's a slightly different message.
```
Error detected while processing /tmp/f.vim[56]..function <SNR>46_F:
line    8:
E1012: Type mismatch; expected object<D> but got object<C>
```
Presumably the result of the `TYPECHECK` that's added by the `<D>`.

**Comment by errael** [comment]
Other than
1. difference between script level and function and `<{type}>`
2. differences in `G1()`, `G2()`, `G3()` in [above example](https://github.com/vim/vim/issues/15432#issuecomment-2272389793)

Are there other issues?

BTW, it's interesting that you think `G1()` is correct and I think it's a bug. It's the difference between `G1()` and `G2()` is what I think is the bug. And disclaimer, as mentioned, I don't particularly like how things currently work, but ...
